<a href="https://colab.research.google.com/github/Majmejia/Business-Inteligence/blob/main/Reto_1_Modelo_NB.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Reto 1** Implementación del modelo Naive Bayes para determinar si una persona vive en un área urbana o rural.

0. Se procede a cargar las librerias de trabajo.

In [ ]:
import numpy as np
import pandas as pd

1. Se procede a cargar la base de datos de Ecxel

In [ ]:
nxl='/content/Reto 1. BD1_Longitudinal Survey_Int.xlsx'
XDB=pd.read_excel(nxl,sheet_name=0)
XDB.head(100)

#Se eliminan los registros que poseen datos faltantes
XDB=XDB.dropna(subset=['Age', 'Mother_Edu', 'Father_Edu', 'Height', 'Weight', 'Income', 'Self_Esteem', 'FamilySize', 'Urban'])

#Se seleccionan las variables de interés
XDB=XDB[['Age', 'Mother_Edu', 'Father_Edu', 'Height', 'Weight', 'Income', 'Self_Esteem', 'FamilySize', 'Urban','Siblings','Black','Hispanic','White','Christian','WomenPlace','Male','Outgoing_Kid','Outgoing_Adult','HealthPlan','Marital_Status','Education','WeeksEmployed','NumberSpouses']]
XDB.head(10)

#Se separan las variables de entrada y de salida
XD=XDB[['Age', 'Mother_Edu', 'Father_Edu','FamilySize', 'Self_Esteem', 'Height', 'Weight', 'Income']] #XD variables de entrada
yd=XDB['Urban'] #Variable de salida


Age  Mother_Edu  Father_Edu  FamilySize  Self_Esteem  Height  Weight  \
1       20         5.0         8.0           5         16.0    62.0   120.0   
5       19        12.0        12.0           4         27.0    64.0   200.0   
6       15        12.0        12.0           3         26.0    65.0   131.0   
7       21         9.0         6.0           3         23.0    65.0   179.0   
8       16        12.0        10.0           6         26.0    66.0   145.0   
...    ...         ...         ...         ...          ...     ...     ...   
12557   20        12.0        12.0           5         20.0    60.0   110.0   
12565   20        12.0        11.0           2         24.0    63.0   152.0   
12647   21        16.0        12.0           1         30.0    70.0   180.0   
12651   21        16.0        16.0           1         27.0    73.0   170.0   
12678   21         8.0         8.0           1         19.0    70.0   180.0   

        Income  
1          0.0  
5      40000.0  
6      25000.0  
7      27400.0  
8      52000.0  
...        ...  
12557   1000.0  
12565  38000.0  
12647  43600.0  
12651  40000.0  
12678  17000.0  

[5948 rows x 8 columns]

2. Se determinan la cantidad total de datos y el número de datos de cada variable

In [ ]:
#Antes de eliminar las filas con datos faltantes.
total_datos1= len(XDB)
print("La cantidad total de datos es:\n",total_datos1)

porcentaje_por_variable1=(XDB.count()/total_datos1)*100
print("\nporcentaje de datos por variable")
print(porcentaje_por_variable1)

#Luego de elimiar las filas con datos faltantes
total_datos= len(XDB)
print(f"Número total de datos: {total_datos}")

porcentaje_por_variable=(XDB.count()/total_datos)*100
print("\nporcentaje de datos por variable")
print(porcentaje_por_variable)

ND1=len(np.where((yd==1))[0])
print(ND1/len(XD)) #Viven en área Urbana

ND2=len(np.where((yd==0))[0])
print(ND2/len(XD)) #Viven en área Rural

La cantidad total de datos es:
 5948

porcentaje de datos por variable
Age               100.000000
Mother_Edu        100.000000
Father_Edu        100.000000
Height            100.000000
Weight            100.000000
Income            100.000000
Self_Esteem       100.000000
FamilySize        100.000000
Urban             100.000000
Siblings           99.899126
Black              98.890383
Hispanic           98.890383
White              98.890383
Christian          99.747814
WomenPlace         99.579691
Male              100.000000
Outgoing_Kid       97.780767
Outgoing_Adult     97.932078
HealthPlan         99.949563
Marital_Status     99.983188
Education         100.000000
WeeksEmployed      99.613315
NumberSpouses     100.000000
dtype: float64
Número total de datos: 5948

porcentaje de datos por variable
Age               100.000000
Mother_Edu        100.000000
Father_Edu        100.000000
Height            100.000000
Weight            100.000000
Income            100.000000
Self_Esteem

# 3. Se procede con la ejecución del modelo Naive Bayes

In [ ]:
from sklearn.naive_bayes import GaussianNB

XD1=XDB[['Age',  'Height', 'Weight']] #Variables de entrada
yd=XDB['Urban'] #Variable de pronóstico

mnb=GaussianNB() #Aloja el modelo
mnb.fit(XD,yd) #Entrena el modelo

#Obtenemos la información del modelo
u=mnb.theta_ #obtenemos la media de las variables Age, Mother_Edu, Father_Edu, Height, Weight, Income, Self_Esteem, FamilySize, Urban (La u representa el valor de la media)
print("La media de 'Age', 'Mother_Edu', 'Father_Edu', 'Height', 'Weight', 'Income', 'Self_Esteem', 'FamilySize', 'Urban':\n",u)

var=mnb.var_ #Se calculan las varianzas
sigma=np.sqrt(var) #Se calcula la desviación
print("La desviación de 'Age', 'Mother_Edu', 'Father_Edu', 'Height', 'Weight', 'Income', 'Self_Esteem', 'FamilySize', 'Urban':\n",sigma)

#Se determinan los limites superiores e inferiores de la Edad, Altura y peso
LS=u+sigma
LI=u-sigma
print("El límite superior de 'Age', 'Mother_Edu', 'Father_Edu', 'Height', 'Weight', 'Income', 'Self_Esteem', 'FamilySize', 'Urban' es:\n",LS)
print("El límite inferior de Age', 'Mother_Edu', 'Father_Edu', 'Height', 'Weight', 'Income', 'Self_Esteem', 'FamilySize', 'Urban' es:\n",LI)

#Se evalua un habitante
XDS=[34,12,15,2,24,69,160,48000]
ydp=mnb.predict([XDS])
print("La predicción del habitante es:\n",ydp)

if ydp==0:
 print("El habitante no vive en área urbana")

if ydp==1:
 print("El habitante vive en área urbana")



La media de 'Age', 'Mother_Edu', 'Father_Edu', 'Height', 'Weight', 'Income', 'Self_Esteem', 'FamilySize', 'Urban':
 [[1.83640032e+01 1.07312349e+01 1.02711864e+01 4.73930589e+00
  2.19951574e+01 6.72429379e+01 1.49003228e+02 3.01248854e+04]
 [1.83618603e+01 1.11613931e+01 1.12331705e+01 4.68103631e+00
  2.25744319e+01 6.69116585e+01 1.44750903e+02 3.32736742e+04]]
La desviación de 'Age', 'Mother_Edu', 'Father_Edu', 'Height', 'Weight', 'Income', 'Self_Esteem', 'FamilySize', 'Urban':
 [[2.47260545e+00 3.15661127e+00 3.88837565e+00 2.55226748e+00
  4.21650807e+00 4.08228636e+00 3.03486350e+01 3.16482374e+04]
 [2.47344359e+00 3.40066236e+00 4.11549881e+00 2.44205851e+00
  4.19763081e+00 4.20193102e+00 3.00285301e+01 3.60897302e+04]]
El límite superior de 'Age', 'Mother_Edu', 'Father_Edu', 'Height', 'Weight', 'Income', 'Self_Esteem', 'FamilySize', 'Urban' es:
 [[2.08366087e+01 1.38878461e+01 1.41595621e+01 7.29157338e+00
  2.62116655e+01 7.13252242e+01 1.79351863e+02 6.17731228e+04]
 [2.083

/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


4. Evaluamos el desempeño del modelo anteriormente hecho.

In [ ]:
from sklearn.metrics import confusion_matrix

#Haga un barrido de los datos
ydp2=mnb.predict(XD) #Haga el pronóstico para los datos base

cm=confusion_matrix(yd,ydp2) #Haga la matriz de confusión
print("La matriz de confusión es:\n",cm)
VN=cm[0,0];FP=cm[0,1];FN=cm[1,0];VP=cm[1,1]

#Metricas Desempeño
Exactitud=(VP+VN)/(VP+VN+FP+FN) #Exactitud: Comportamiento general preaprobando
print("La exactitud del modelo es:\n",Exactitud)

TE=(FN+FP)/(VP+VN+FP+FN) #Tasa de Error: Comportamiento general no preaprobando
print("La tasa de error del modelo es:\n",TE)

Sensibilidad=VP/(VP+FN) #Sensibilidad: Comportamiento correcto preaprobando
                        #Sobre el total de los preaprobandos
print("La sensibilidad del modelo es:\n",Sensibilidad)

Sp=VN/(VN+FP) #Especificidad: Comportamiento correcto no preaprobando
              #Sobre el total de los no preaprobandos
print("La especificidad del modelo es:\n",Sp)

Pr=VP/(VP+FP) #Precisión: Comportamiento correcto preaprobando
             #Sobre el total de los preaprobados
print("La precisión del modelo es:\n",Pr)

PrN=VN/(VN+FN) #Precisión Negativa: Comportamiento correcto no preaprobando
             #Sobre el total de los no preaprobados
print("La precisión negativa del modelo es:\n",PrN)

La matriz de confusión es:
 [[   3 1236]
 [   8 4701]]
La exactitud del modelo es:
 0.7908540685944856
La tasa de error del modelo es:
 0.20914593140551446
La sensibilidad del modelo es:
 0.9983011255043533
La especificidad del modelo es:
 0.002421307506053269
La precisión del modelo es:
 0.7918140474987367
La precisión negativa del modelo es:
 0.2727272727272727


**Análisis de resultados**
La base de datos suministrada se tomo como referencia para lograr determinar si una persona vive en un área urbana o rural. La cantidad total de lo datos es de (5948). Al poner a prueba al modelo se puede observar que este alcanzo una exactitud del 79% aproximadamente, demostrando un buen comportamiento para determinar si una persona vive en el área urbana o rural. Además de esto, el índice de precisión alcanzo un valor cercano al 80% en promedio (79.18%) lo que demuestra un criterio aceptable de el modelo para determinar si una persona reside en una zona urbana o rural.